File: dhts/.deepcode/dc_config.py
"""DeepCode shared config loader for Python tools.

Discovers and merges settings from (in priority order, last wins):
  1. ~/.deepcode/settings.json   (user-global)
  2. ./.deepcode/settings.json   (project-local)
  3. env var DEEPCODE_SETTINGS   (comma-separated extra paths)

Usage:
    from dc_config import load_settings
    settings = load_settings()
    print(settings["workspace"]["name"])
"""

from __future__ import annotations

import json
import os
from pathlib import Path
from typing import Any


def _find_project_root(start: Path | None = None) -> Path:
    """Walk upward from *start* looking for a .deepcode/ directory."""
    candidate = (start or Path.cwd()).resolve()
    for ancestor in [candidate, *candidate.parents]:
        if (ancestor / ".deepcode" / "settings.json").exists():
            return ancestor
    return candidate


def _merge_dicts(base: dict[str, Any], overlay: dict[str, Any]) -> dict[str, Any]:
    """Deep-merge *overlay* into *base*. Lists are replaced, not appended."""
    for key, value in overlay.items():
        if key in base and isinstance(base[key], dict) and isinstance(value, dict):
            base[key] = _merge_dicts(base[key], value)
        else:
            base[key] = value
    return base


def load_settings(
    start: Path | None = None,
    extra_paths: list[str] | None = None,
) -> dict[str, Any]:
    """Return the merged DeepCode settings dictionary.

    *start* – directory to begin the project-root search.
    *extra_paths* – additional JSON paths to merge (highest priority).
    """
    settings: dict[str, Any] = {}

    # 1) user-global
    global_path = Path.home() / ".deepcode" / "settings.json"
    if global_path.exists():
        try:
            settings = _merge_dicts(settings, json.loads(global_path.read_text("utf-8")))
        except (json.JSONDecodeError, OSError):
            pass

    # 2) project-local
    project_root = _find_project_root(start)
    project_path = project_root / ".deepcode" / "settings.json"
    if project_path.exists():
        try:
            settings = _merge_dicts(settings, json.loads(project_path.read_text("utf-8")))
        except (json.JSONDecodeError, OSError):
            pass

    # 3) DEEPCODE_SETTINGS env var
    env_paths = os.environ.get("DEEPCODE_SETTINGS", "")
    for p in env_paths.split(","):
        p = p.strip()
        if p and Path(p).exists():
            try:
                settings = _merge_dicts(settings, json.loads(Path(p).read_text("utf-8")))
            except (json.JSONDecodeError, OSError):
                pass

    # 4) extra_paths (caller-provided)
    for p in (extra_paths or []):
        if p and Path(p).exists():
            try:
                settings = _merge_dicts(settings, json.loads(Path(p).read_text("utf-8")))
            except (json.JSONDecodeError, OSError):
                pass

    return settings


def resolve_path(settings: dict[str, Any], key: str) -> Path:
    """Resolve a settings path relative to the dhts root.

    If the key ends with 'Dir', ensure the directory exists.
    """
    paths = settings.get("paths", {})
    raw = paths.get(key, "")
    if not raw:
        raise KeyError(f"paths.{key} not found in settings")
    dhts_root = Path(paths.get("dhtsRoot", "."))
    resolved = (dhts_root / raw).resolve()
    return resolved

File: dhts/.deepcode/dc_config.ts
/**
 * DeepCode shared config loader for TypeScript tools.
 *
 * Discovers and merges settings from (in priority order, last wins):
 *   1. ~/.deepcode/settings.json   (user-global)
 *   2. ./.deepcode/settings.json   (project-local)
 *   3. env var DEEPCODE_SETTINGS   (comma-separated extra paths)
 *
 * Usage:
 *   import { loadSettings } from '../../.deepcode/dc_config.js';
 *   const settings = loadSettings();
 */

import { readFileSync } from 'node:fs';
import { homedir } from 'node:os';
import { join, resolve, dirname } from 'node:path';

export interface DeepCodeSettings {
  version: string;
  description: string;
  workspace: {
    name: string;
    root: string;
    company: string;
    program: string;
  };
  paths: Record<string, string>;
  scan: {
    excludeDirectories: string[];
    includeRoots: string[];
    sourceExtensions: string[];
    markdownExtensions: string[];
  };
  execution: {
    timeoutSeconds: number;
    maxConcurrency: number;
    httpTimeoutSeconds: number;
    retryAttempts: number;
  };
  wip: {
    maxActiveGlobal: number;
    maxActivePerLane: number;
  };
  ai: {
    proxyPort: number;
    defaultProvider: string;
    failoverProvider: string;
    budgetAlertThreshold: number;
  };
  cadence: Record<string, string[]>;
  pipelines: Record<string, string[]>;
  qualityGates: Record<string, { description: string; workflows: string[] }>;
  workflows: Record<string, { cwd: string; command: string }>;
  [key: string]: unknown;
}

function deepMerge<T extends Record<string, unknown>>(
  target: T,
  source: Record<string, unknown>,
): T {
  for (const key of Object.keys(source)) {
    if (
      key in target &&
      typeof target[key] === 'object' &&
      !Array.isArray(target[key]) &&
      typeof source[key] === 'object' &&
      !Array.isArray(source[key])
    ) {
      target[key] = deepMerge(
        target[key] as Record<string, unknown>,
        source[key] as Record<string, unknown>,
      ) as T[Extract<keyof T, string>];
    } else {
      (target as Record<string, unknown>)[key] = source[key];
    }
  }
  return target;
}

function tryReadJson(path: string): Record<string, unknown> | null {
  try {
    return JSON.parse(readFileSync(path, 'utf-8'));
  } catch {
    return null;
  }
}

function findProjectRoot(start?: string): string {
  let dir = start ? resolve(start) : process.cwd();
  // eslint-disable-next-line no-constant-condition
  while (true) {
    try {
      readFileSync(join(dir, '.deepcode', 'settings.json'));
      return dir;
    } catch {
      /* not here */
    }
    const parent = dirname(dir);
    if (parent === dir) break;
    dir = parent;
  }
  return process.cwd();
}

export function loadSettings(startDir?: string): DeepCodeSettings {
  let settings: Record<string, unknown> = {};

  // 1) user-global
  const globalPath = join(homedir(), '.deepcode', 'settings.json');
  const global = tryReadJson(globalPath);
  if (global) settings = deepMerge(settings, global);

  // 2) project-local
  const projectRoot = findProjectRoot(startDir);
  const projectPath = join(projectRoot, '.deepcode', 'settings.json');
  const project = tryReadJson(projectPath);
  if (project) settings = deepMerge(settings, project);

  // 3) DEEPCODE_SETTINGS env var
  const envPaths = process.env.DEEPCODE_SETTINGS || '';
  for (const p of envPaths.split(',')) {
    const trimmed = p.trim();
    if (trimmed) {
      const extra = tryReadJson(trimmed);
      if (extra) settings = deepMerge(settings, extra);
    }
  }

  return settings as unknown as DeepCodeSettings;
}

export function resolvePath(settings: DeepCodeSettings, key: string): string {
  const paths = settings.paths || {};
  const raw = paths[key];
  if (!raw) throw new Error(`paths.${key} not found in settings`);
  const dhtsRoot = paths.dhtsRoot || '.';
  return resolve(dhtsRoot, raw);
}
